### PR TITLE
feat: Introduce AutoUseCompleter, now support def

### PR DESCRIPTION
**Pull request by stormckey**
first step for #9347.

Preview:
image
image
image

Currently we use `AutoUseCompletion` to provide the option, but we can actually make the completion option more informative if we design a specific completion for it.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I will merge this, but can we not provide more information like the DefCompleter does?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala [hunk 1]
```diff
@@ -562,6 +572,17 @@ object Completion {
...
+  /**
+   * Represents an auto-import completion.
+   *
+   * @param label         the label of the completion, displayed in the completion item.
+   * @param name          the name to be completed under cursor.
+   * @param path          the path of a flix construct to be used.
+   * @param ap            the anchor position for the use statement.
+   * @param kind          the kind of the completion.
+   */
+  case class AutoUseCompletion(label: String, name:String, path: String, ap: AnchorPosition, kind: CompletionItemKind) extends Completion
```
[the remark this answers is not in the document]
Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:584]: Maybe renamed it to qualifiedName? It's something like `A.bar`.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ExprCompleter.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ import ca.uwaterloo.flix.language.ast.TypedAst
 object ExprCompleter {
 
   def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root): Iterable[Completion] = {
-    DefCompleter.getCompletions(context) ++
+//    DefCompleter.getCompletions(context) ++
```
[the remark this answers is not in the document]
Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ExprCompleter.scala:26]: The duty of this part will later be taken by the `Declaration` part in `LocalScopeCompletion`.

When it's not in scope, `AutoUseCompletion` will provide the choice. Otherwise, `LocalScopeCompletion` will take it.

This is also part of moving SyntacticCompletion to SemanticCompletion.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        val name = decl.sym.name
+        val label = s"$name (use $path)"
+        Completion.AutoUseCompletion(label, name, path, ap, Function)
+      }
+  }
+
+  /**
+    * Returns `true` if the given definition `decl` is not in the given `scope`.
+    */
+  private def outOfScope(decl: TypedAst.Def, scope: LocalScope): Boolean = {
```
[the remark this answers is not in the document]
Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala:69]: Emmm what does that means? I write this function to make sure `AutoUseCompleter` will keep silent for things that are already in scope, when it't time for `LocalScopeCompleter` to do its job.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completion.scala:584]: Yes. Qualified is a better name.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala:69]: I meant, do we not have a check like this already in the AutoImporter?

**Comment by stormckey** [comment]
> I will merge this, but can we not provide more information like the DefCompleter does?

Yes we can, and we will do so by using a dedicated completion for it.

Initially it will just be something like DefCompletion + extra `use ...`.

Maybe later we can discuss to make the extra `use ...` just an optional field for `DefCompletion` .

**Comment by magnus-madsen** [comment]
OK; if you fix the smaller things then I can merge and we can iterate.

**Comment by stormckey** [comment]
Well now it looks like this:
image
with snippet support:
image

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ExprCompleter.scala [hunk 4]
```diff
@@ -23,7 +23,6 @@ import ca.uwaterloo.flix.language.ast.TypedAst
 object ExprCompleter {
 
   def getCompletions(context: CompletionContext)(implicit flix: Flix, index: Index, root: TypedAst.Root): Iterable[Completion] = {
-    DefCompleter.getCompletions(context) ++
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/ExprCompleter.scala:26]: I am still confused why you remove this.

If you remove it then I can no longer type `List.m` and have it completed to `List.map`?

**Comment by magnus-madsen** [comment]
What does it look like if you type `List.m`?

**Comment by stormckey** [comment]
> What does it look like if you type `List.m`?
This:
image

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala [hunk 5]
```diff
@@ -46,7 +46,7 @@ object DefCompleter {
     val isMatch = if (isNamespace)
       decl.sym.toString.startsWith(word)
     else
-      decl.sym.text.startsWith(word)
+      false
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala:49]: Why this change?

Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala:49]: Otherwise, DefCompleter will provide `A.bar` for `ba` when there is already `use A.bar;`

Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala:49]: All possible cases now:
List.ma => DefCompleter
ma (without `use List.map;`) => AutoUseCompleter
ma (with `use List.map`) => LocalScopeCompleter

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    *
+    *  We propose to complete the name to `bar` and use the function `A.bar`
+    *
+    *  {{{
+    *    use A.bar;
+    *    ...
+    *    let s = bar
+    *  }}}
+    */
+  def getCompletions(err: ResolutionError.UndefinedName, ctx: CompletionContext)(implicit root: TypedAst.Root): Iterable[Completion] =
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala:48]: You are not allowed to use CompletionContext in a semantic completer.

In particular, you should NOT use .word. 

Instead, you can get the name from the error.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala (hunk 5 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/DefCompleter.scala:49]: OK- that makes sense.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/completion/AutoUseCompleter.scala:66]: No, and the reason is in AutoImportCompleter, we can just check the keys in `LocalScope` against class names from `AvailableClasses`. But there, the key is just an unqualified `bar`, which is not what we want. Additionally, we have to extract the qualified name of the candidate function, which is also different from the handy `AvailableClasses` in `AutoImportCompleter`
